filebeat/inputsource/udp/server.go

```diff
@@ -0,0 +1,79 @@
+package udp
+
+import (
+	"net"
+	"time"
+
+	"github.com/elastic/beats/libbeat/common/atomic"
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+// Config options for the UDPServer
+type Config struct {
+	Host           string        `config:"host"`
+	MaxMessageSize int           `config:"max_message_size"`
+	Timeout        time.Duration `config:"timeout"`
+}
+
+// Server creates a simple UDP Server and listen to a specific host:port and will send any
+// event received to the callback method.
+type Server struct {
+	running  atomic.Bool
+	config   *Config
+	callback func(data []byte)
+	listener net.PacketConn
+}
+
+// New returns a new UDPServer instance.
+func New(config *Config, callback func(data []byte)) *Server {
+	return &Server{
+		running:  atomic.MakeBool(false),
+		config:   config,
+		callback: callback,
+	}
+}
+
+// Run starts the UDP Server and listen to incoming events.
+func (u *Server) Run() error {
+	var err error
+
+	u.listener, err = net.ListenPacket("udp", u.config.Host)
+	if err != nil {
+		return err
+	}
+	defer u.listener.Close()
+	u.running.Swap(true)
+
+	logp.Info("Started listening for UDP on: %s", u.config.Host)
+
+	buffer := make([]byte, u.config.MaxMessageSize)
+
+	for u.running.Load() {
+		u.listener.SetDeadline(time.Now().Add(u.config.Timeout))
+		length, _, err := u.listener.ReadFrom(buffer)
+		if err != nil {
+			// don't log any deadline events.
+			e, ok := err.(net.Error)
+			if !ok && !e.Timeout() {
+				logp.Err("Error reading from buffer: %v", err.Error())
+			}
+			continue
+		}
+
+		if length > 0 {
+			u.callback(buffer[:length])
+		}
+	}
+	return nil
+}
+
+//IsRunning returns true when the UDP server is accepting new data.
+func (u *Server) IsRunning() bool {
+	return u.running.Load()
+}
+
+// Stop stops the current udp server.
+func (u *Server) Stop() {
+	logp.Info("Stopping UDP server on: %s", u.config.Host)
+	u.running.Swap(false)
```

Comment:
I'd prefer for `Stop` to block until the goroutine exits. This should `Close` the socket and wait for `run` to exit (via `sync.WaitGroup`).

Signaling to stop via a flag means that the socket won't actually be closed until the a packet is received to unblock the recvfrom or a timeout is reached (max 5 minutes).